lfs.h

```diff
@@ -219,6 +220,99 @@ struct lfs_config {
     // LFS_ATTR_MAX when zero.
     lfs_size_t attr_max;
 };
+#else
+// Static configuration if LFS_STATICCFG is defined, there are defaults
+// for some of these, but some are required. For full documentation, see
+// the lfs_config struct above.
+
+// Block device operations
+int lfs_read(lfs_block_t block,
+        lfs_off_t off, void *buffer, lfs_size_t size);
+int lfs_prog(lfs_block_t block,
+        lfs_off_t off, const void *buffer, lfs_size_t size);
+int lfs_erase(lfs_block_t block);
+int lfs_sync(void);
+
+// Required configuration
+#ifndef LFS_READ_SIZE
+#error "LFS_STATICCFG requires LFS_READ_SIZE"
+#endif
+#ifndef LFS_PROG_SIZE
+#error "LFS_STATICCFG requires LFS_PROG_SIZE"
+#endif
+#ifndef LFS_BLOCK_SIZE
+#error "LFS_STATICCFG requires LFS_BLOCK_SIZE"
+#endif
+#ifndef LFS_BLOCK_COUNT
+#error "LFS_STATICCFG requires LFS_BLOCK_COUNT"
+#endif
+#ifndef LFS_BLOCK_CYCLES
+#error "LFS_STATICCFG requires LFS_BLOCK_CYCLES"
+#endif
+#ifndef LFS_CACHE_SIZE
+#error "LFS_STATICCFG requires LFS_CACHE_SIZE"
+#endif
+#ifndef LFS_LOOKAHEAD_SIZE
+#error "LFS_STATICCFG requires LFS_LOOKAHEAD_SIZE"
+#endif
```

Comment:
Food for thought, but another pattern could have been something like:

```
struct lfs_config
{
...
#ifndef LFS_STATICCFG_READ_SIZE
    lfs_size_t read_size;
#endif
#ifndef LFS_STATICCFG_WRITE_SIZE
...
};
...
static inline lfs_size_t lfs_cfg_read_size(const struct lfs_config *c)
{
#ifndef LFS_STATICCFG_READ_SIZE
    return c->read_size;
#else
    return LFS_STATICCFG_READ_SIZE;
#endif
}
```

That would allow for:
- Partially optimized builds for any configuration.  E.g. a user could declare what variables they'd never change for a given static configuration (but maybe would change a different parameter), and this could account for that.
- You could add min/max checks for the dynamic and/or static configurations that are used across the board.  For instance, the block cycles need to be either off, or at least 2 IIRC.
- Having the functions in the header (not strictly required) allows a single point of entry for users to read a given configuration item regardless of if it's static or dynamic.    Right now all of the helpers are placed into the lfs.c file, but it'd be potentially useful for a single point of retrieval.

Reply:
(Granted, doing the above, would require a cfg object to be always passed around to all functions instead of being able to remove it as in this patch; but this could get optimized out with link time optimization.)

Reply:
Hmm, I'll have to see what this looks like. Partial optimizations would be nice, though it may be a bit difficult to track down all constant-related warnings.

> Granted, doing the above, would require a cfg object to be always passed around to all functions

We could always do a big:
``` c
#ifdef LFS_STATICCFG_READ_SIZE && \
        LFS_STATICCFG_PROG_SIZE && \
        LFS_STATICCFG_BLOCK_SIZE && \
        LFS_STATICCFG_BLOCK_COUNT && \
        ...
```

> You could add min/max checks for the dynamic and/or static configurations that are used across the board. For instance, the block cycles need to be either off, or at least 2 IIRC.

Currently these are checked by runtime asserts in `lfs_init`. Unfortunately the math isn't always decidable at compile-time by C, but the runtime checks can be removed if you define `LFS_NO_ASSERTS`.

Reply:
I went ahead and created something since I was curious.  See here: https://github.com/tim-nordell/littlefs/commit/734de59558a005db32589a5efbd5d6591d4a3828

Reply:
Interestingly, if the avoidance of division in the eviction logic is applied (#489) in conjunction with a static configuration using powers of 2 for all parameters, we end up with only 1 function performing division with a static configuration: `lfs_ctz_index(...)`

That function does a non div-by-power-of-2 since it divides by the block size less 2 * 4.  It looks like that function calculates something integral to the design of LFS, so the - (2 * 4) can't be removed (to optimize for architectures without a hardware divider).

Reply:
Oh! That is an interesting observation.

It doesn't surprise me that `lfs_ctz_index` gets in the way, it is the most complex math in littlefs. The division is effectively finding the number of blocks to iterate over, but each block has on average 2 pointers, so it doesn't work out that nicely.

Hmm, but the block_size is likely a power-of-two, and 2*4 is a power of two. Is there a shortcut for division by the sum of two powers of two?

Reply:
Pushed up a change to do this 👍 I believe the main difference is I'm using macro names without the LFS_STATICCFG_* prefix.

Reply:
:) Glad you liked the idea of partials.

Note that using MACROs like that loses the type safetyness of using the static inline functions, but it's probably just a stylistic thing here.  There shouldn't be a difference in the compiled size between the two choices if the compiler is doing its job.

Also, I don't know what target C standard this is targeted at (C99, C11, etc), but if it was C11, you can use static asserts (https://en.cppreference.com/w/c/language/_Static_assert) for some of the configuration values statically specified.

Reply:
> Note that using MACROs like that loses the type safetyness of using the static inline functions

True, though in this case they are just rearranging arguments for another function. Also, if all of the macros are specified, we wouldn't emit the `lfs_cfg` struct, which would cause a type error if these were inline functions.

> Also, I don't know what target C standard this is targeted at (C99, C11, etc), but if it was C11, you can use static asserts

Ah, littlefs sticks to C99, and even this has been a pain point for users. Embedded IDEs move pretty slowly.

Though one nice thing with static_asserts is that you can always have a macro that switches between nothing or a static assert if available. 

The main downside would be figuring out what can be static asserted, `LFS_ASSERT(LFS_CFG_BLOCK_SIZE(cfg) > LFS_CFG_PROG_SIZE(cfg))` may or may not be static assertable depending on what's defined. And it risks duplicating a lot of these checks.